File: euler/250/250.cc
#include <cstdio>
#include <cstring>
#include <algorithm>

typedef long long ll;
const ll MOD = (ll)(1e16);
const int N = 250;

int n;
ll dp[N], tmp[N];

int fastpow(int base, int pow, int mod) {
  if (!pow) return 1;
  ll tmp = fastpow(base, pow>>1, mod);
  return (pow & 1) ? (tmp * tmp * base)%mod : (tmp * tmp)%mod;
}

int main() {
  scanf("%d", &n);
  //printf("%d\n", fastpow(2, n, N));
  memset(dp, 0, sizeof dp);
  for (int i=1; i<=n; i++) {
    int val = fastpow(i, i, N);
    memcpy(tmp, dp, sizeof dp);
    for (int j=0; j<N; j++) {
      dp[j] = (dp[j] + tmp[(j + N - val) % N]) % MOD;
    }
    dp[val] = (dp[val] + 1) % MOD;
  }
  for (int i=0; i<N; i++) {
    printf("dp[%d] = %lld\n", i, dp[i]);
  }

  return 0;
}

File: euler/250/250.py
#!/usr/bin/python
MOD = 10**16
MAX = 250250
def fastpow(a, b, mod):
   if b == 0:
      return 1
   tmp = fastpow(a, b/2, mod)
   ans = (tmp * tmp ) % mod;
   if (b % 2 == 1):
      ans = (ans * a) % mod;
   return ans

dp = [0] * 251;
tmp = [0] * 251;
for k in range(1, MAX+1):
   n = fastpow(k, k, 250)
   for i in range(0, 250):
       tmp[i] = dp[i]
   for i in range(0, 250):
       dp[i] = (dp[i] + tmp[(i-n+250)%250]) % MOD
   dp[n] = (dp[n] + 1) % MOD

print dp[0]
